docs/madr/decisions/046-grpc-probes.md

```diff
@@ -0,0 +1,239 @@
+  name: busybox
+  labels:
+    run: busybox
+  annotations:
+    kuma.io/virtual-probes-port: "19000"
+spec:
+  containers:
+    - name: busybox
+      image: busybox
+      resources: {}
+      readinessProbe:
+        httpGet:
+          path: /metrics
+          port: 3001
+        initialDelaySeconds: 3
+        periodSeconds: 3
+      livenessProbe:
+        httpGet:
+          path: /metrics
+          port: 8080
+        initialDelaySeconds: 3
+        periodSeconds: 3
+```
+
+Pod with sidecar injected:
+
+[inject.14.golden.yaml](https://github.com/kumahq/kuma/blob/master/pkg/plugins/runtime/k8s/webhooks/injector/testdata/inject.14.golden.yaml)
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  annotations:
+    kuma.io/virtual-probes: enabled
+    kuma.io/virtual-probes-port: "19000"
+    ...
+  labels:
+    run: busybox
+  name: busybox
+spec:
+  containers:
+  - name: kuma-sidecar
+    image: kuma/kuma-sidecar:latest
+    ...
+  - image: busybox
+    livenessProbe:
+      httpGet:
+        path: /8080/metrics
+        port: 19000
+      initialDelaySeconds: 3
+      periodSeconds: 3
+    name: busybox
+    readinessProbe:
+      httpGet:
+        path: /3001/metrics
+        port: 19000
+      initialDelaySeconds: 3
+      periodSeconds: 3
+  initContainers:
+  - name: kuma-init
+    image: kuma/kuma-init:latest
+  ...
+```
+
+And in the converted `Dataplane` object, we can see the two probes are both handled by the same Virtual Probe listener on port 19000:
+
+```yaml
+type: Dataplane
+mesh: default
+metadata:
+  creationTimestamp: null
+spec:
+  probes:
+    endpoints:
+      - inboundPath: /metrics
+        inboundPort: 8080
+        path: /8080/metrics
+      - inboundPath: /metrics
+        inboundPort: 3001
+        path: /3001/metrics
+    port: 19000
+  ...
+```
+
+## Considered Options
+
+**Option 1:** Transform gRPC probes to HTTP probes, merging all the HTTP probes into the existing HTTP based Virtual Probes listener
+
+**Option 2:** Allocate a new and separated port for each of the gRPC probes, and route them back to the application probe handling port accordingly. 
+
+### Proposed option: option 2 - allocate a new and separated port for each of the gRPC probes, and route them back to the application probe handling port accordingly 
+
+The only information we can use to differ from multiple gRPC probes is the port number, while the port number is included in the underlying HTTP2 request as `Host` header, it will be the same if we want to use a single virtual probe listener to support forwarding traffic for all gRPC probes. So we need to allocate a new virtual probe port for each of them. To allocate these ports, we'll need a new Pod annotation to support user specifying the range to be used. This annotation, with name `kuma.io/virtual-probes-port-range`, will replace the existing annotation `kuma.io/virtual-probes-port`. The annotation should be optional: when user does not specify, a default range will be used.
```

Comment:
multiple gRPC probes to the same service no? Because otherwise the service enables us to make a difference between probes no?

Reply:
What is this range?

Reply:
> multiple gRPC probes to the same service no? Because otherwise the service enables us to make a difference between probes no?

the `service` value will actually be carried over the gRPC payload, insteand of any HTTP header, so it's not available for routing.

Reply:
I'll add the proposed range.